DeepSpeech.py

```diff
@@ -447,61 +447,74 @@ def train():
                           ' - consider using load option "auto" or "init".' % FLAGS.load)
                 sys.exit(1)
 
-        def run_set(set_name, init_op, num_batches):
+        def run_set(set_name, init_op):
             is_train = set_name == 'train'
             train_op = apply_gradient_op if is_train else []
             feed_dict = dropout_feed_dict if is_train else no_dropout_feed_dict
+
             total_loss = 0.0
+            step_count = 0
+
             step_summary_writer = step_summary_writers.get(set_name)
-            num_steps = max(1, num_batches // len(Config.available_devices))
             checkpoint_time = time.time()
 
+            class LossWidget(progressbar.widgets.FormatLabel):
+                def __init__(self):
+                    progressbar.widgets.FormatLabel.__init__(self, format='Loss: %(mean_loss)f')
+
+                def __call__(self, progress, data):
+                    data['mean_loss'] = total_loss / step_count if step_count else 0.0
+                    return progressbar.widgets.FormatLabel.__call__(self, progress, data)
+
             if FLAGS.show_progressbar:
-                pbar = progressbar.ProgressBar(max_value=num_steps, redirect_stdout=True).start()
-            else:
-                pbar = lambda i: i
+                pbar = progressbar.ProgressBar(widgets=['Epoch {}'.format(epoch),
+                                                        ' | ', progressbar.widgets.Timer(),
+                                                        ' | Steps: ', progressbar.widgets.Counter(),
+                                                        ' | ', LossWidget()])
+                pbar.start()
 
             # Initialize iterator to the appropriate dataset
             session.run(init_op)
 
             # Batch loop
-            for step_index in pbar(range(num_steps)):
-                if coord.should_stop():
+            while True:
+                try:
+                    _, current_step, batch_loss, step_summary = \
+                        session.run([train_op, global_step, loss, step_summaries_op],
+                                    feed_dict=feed_dict)
+                except tf.errors.OutOfRangeError:
```

Comment:
"You shall not rely on exception handling for program flow control!" - Honestly: I am totally OK with it. Actually I like it.

Reply:
Yeah, it's unfortunate. When we have eager execution this will be:

```python
for batch in dataset:
    # ...
```